lib/Widgets/Utils.vala

```diff
@@ -204,6 +204,68 @@ public static string markup_accel_tooltip (string[]? accels, string? description
     return string.joinv ("\n", parts);
 }
 
+private double contrast_ratio (Gdk.RGBA bg_color, Gdk.RGBA fg_color) {
+    var bg_luminance = get_luminance (bg_color);
+    var fg_luminance = get_luminance (fg_color);
+
+    if (bg_luminance > fg_luminance) {
+        return (bg_luminance + 0.05) / (fg_luminance + 0.05);
+    }
+
+    return (fg_luminance + 0.05) / (bg_luminance + 0.05);
+}
+
+private double get_luminance (Gdk.RGBA color) {
+    var red = sanitize_color (color.red) * 0.2126;
+    var green = sanitize_color (color.green) * 0.7152;
+    var blue = sanitize_color (color.blue) * 0.0722;
+
+    return (red + green + blue);
+}
+
+private double sanitize_color (double color) {
+    if (color <= 0.03928) {
+        return color / 12.92;
+    }
+
+    return Math.pow ((color + 0.055) / 1.055, 2.4);
+}
+
+/**
+ * Takes a {@link Gdk.RGBA} background color and returns a suitably-contrasting foreground color, i.e. for determining text color on a colored background.
+ *
+ * @param bg_color a {@link Gdk.RGBA} background color
+ *
+ * @return a contrasting {@link Gdk.RGBA} foreground color
```

Comment:
This should probably also mention that the only colors that can be returned is either completely black or completely white and possible values e.g:

> The return value can be either a white color (`{ 1.0, 1.0, 1.0, 1.0`}, or a black color (`{ 0.0, 0.0, 0.0, 0.0}`) depending on the contrast of the passed {@link Gdk.RGBA} color.

Reply:
@donadigo that's true now, but I was wondering if we'd want to do anything more clever with shades of the background color, or not. So I was leaving it open.